enzyme/Enzyme/MLIR/Passes/PrintAliasAnalysis.cpp

```diff
@@ -0,0 +1,129 @@
+    solver.load<dataflow::SparseConstantPropagation>();
+
+    if (failed(solver.initializeAndRun(getOperation()))) {
+      return signalPassFailure();
+    }
+    raw_ostream &os = llvm::outs();
+
+    SmallVector<std::pair<Attribute, Value>> taggedPointers;
+    getOperation()->walk([&](FunctionOpInterface funcOp) {
+      for (auto arg : funcOp.getArguments()) {
+        if (funcOp.getArgAttr(arg.getArgNumber(), "enzyme.tag")) {
+          taggedPointers.push_back(
+              {funcOp.getArgAttr(arg.getArgNumber(), "enzyme.tag"), arg});
+        }
+      }
+
+      funcOp.walk([&](Operation *op) {
+        if (op->hasAttr("tag")) {
+          for (OpResult result : op->getResults()) {
+            taggedPointers.push_back({op->getAttr("tag"), result});
+          }
+        }
+      });
+    });
+
+    for (const auto &[tag, value] : taggedPointers) {
+      const auto *state = solver.lookupState<enzyme::AliasClassLattice>(value);
+      if (state) {
+        os << "tag " << tag << " " << *state << "\n";
+      }
+    }
+
+    if (taggedPointers.empty())
+      return;
+
+    // Compare all tagged pointers
+    for (unsigned i = 0; i < taggedPointers.size() - 1; i++) {
+      for (unsigned j = i + 1; j < taggedPointers.size(); j++) {
+        const auto &[tagA, a] = taggedPointers[i];
+        const auto &[tagB, b] = taggedPointers[j];
+
+        const auto *lhs = solver.lookupState<enzyme::AliasClassLattice>(a);
+        const auto *rhs = solver.lookupState<enzyme::AliasClassLattice>(b);
+        if (!(lhs && rhs))
+          continue;
+
+        os << tagA << " and " << tagB << ": " << lhs->alias(*rhs) << "\n";
+      }
+    }
+
+    getOperation()->walk([&solver, &os](Operation *op) {
+      if (auto funcOp = dyn_cast<FunctionOpInterface>(op)) {
+        for (auto arg : funcOp.getArguments()) {
+          auto *state = solver.lookupState<enzyme::AliasClassLattice>(arg);
+          if (state) {
+            for (auto aliasClass : state->getAliasClasses())
+              funcOp.setArgAttr(arg.getArgNumber(), "enzyme.ac", aliasClass);
+          }
+        }
+      } else if (op->hasTrait<OpTrait::ReturnLike>() &&
+                 isa<FunctionOpInterface>(op->getParentOp())) {
+        os << "points-to-pointer sets for op @" << op->getLoc() << ":\n";
+        if (auto *state = solver.lookupState<enzyme::PointsToSets>(op))
+          os << *state << "\n";
+        else
+          os << "NOT computed\n";
+      }
+      if (op->hasAttr("tag")) {
+        for (OpResult result : op->getResults()) {
+          auto *state = solver.lookupState<enzyme::AliasClassLattice>(result);
+          if (state) {
+            if (state->isUnknown()) {
+              op->setAttr("ac",
+                          StringAttr::get(result.getContext(), "<unknown>"));
+            } else {
+              for (auto aliasClass : state->getAliasClasses()) {
+                op->setAttr("ac", aliasClass);
+              }
```

Comment:
This overwrites the attribute for every new result, should use an array attribute instead with as many entries as the op has results.